pkg/distributor/ha_tracker.go

```diff
@@ -453,20 +455,27 @@ func (h *defaultHaTracker) updateKVStoreAll(ctx context.Context, now time.Time)
 			}
 			var replica string
 			var receivedAt int64
+			var sampleTime time.Time
 			if uh.withinUpdateTimeout(now, entry.electedLastSeenTimestamp) {
 				// We have seen the elected replica recently; carry on with that choice.
 				replica = entry.elected.Replica
 				receivedAt = entry.electedLastSeenTimestamp
+				sampleTime = now
 			} else if uh.withinUpdateTimeout(now, entry.nonElectedLastSeenTimestamp) {
 				// Not seen elected but have seen another: attempt to fail over.
 				replica = entry.nonElectedLastSeenReplica
 				receivedAt = entry.nonElectedLastSeenTimestamp
+				if uh.useSampleTimeForFailover {
+					sampleTime = timestamp.Time(entry.nonElectedLastSeenSampleTime)
+				} else {
+					sampleTime = now
+				}
 			} else {
 				continue // we don't have any recent timestamps
 			}
 			// Release lock while we talk to KVStore, which could take a while.
 			h.electedLock.RUnlock()
-			err := uh.updateKVStore(ctx, cluster, replica, now, receivedAt)
+			err := uh.updateKVStore(ctx, cluster, replica, now, sampleTime, receivedAt, true)
```

Comment:
why pass both times to updateKVStore? can't we change between the time of the sample and now

Reply:
Not sure if I understand your question, are you asking if we can check for the new config flag, and then pass `now`/`sampleTime` but not both, without changing the signature? If so:

Within `updateKVStore` we will always use the `now` value, as it is used in multiple places, e.g. for `electedAt` time and to check failover. If the new flag is set to true, we will use `sampleTime` to check failover, but we will still keep `now` for `electedAt`, to avoid changing more than is necessary for the fix.